gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageImpl.java

```diff
@@ -2248,6 +2275,67 @@ private boolean canIgnoreExceptionForEmptyObject(
     return false;
   }
 
+  @Override
+  public boolean isHnBucket(URI src) throws IOException {
+    String bucketName = src.getAuthority();
+    Boolean isEnabled = cache.getIfPresent(bucketName);
+    if (isEnabled != null) {
+      return isEnabled;
+    }
+
+    String prefix = src.getPath().substring(1);
+
+    StorageControlClient storageControlClient = lazyGetStorageControlClient();
+    GetStorageLayoutRequest request =
+        GetStorageLayoutRequest.newBuilder()
+            .setPrefix(prefix)
+            .setName(StorageLayoutName.format("_", bucketName))
+            .build();
+
+    try (ITraceOperation to = TraceOperation.addToExistingTrace("getStorageLayout.Hn")) {
+      boolean result =
+          storageControlClient.getStorageLayout(request).getHierarchicalNamespace().getEnabled();
+
+      logger.atFine().log("Checking if %s is HN enabled returned %s", src, result);
+
+      cache.put(bucketName, result);
+
+      return result;
+    }
+  }
+
+  private StorageControlClient lazyGetStorageControlClient() throws IOException {
+    if (this.storageControlClient == null) {
+      this.storageControlClient =
+          StorageControlClient.create(
+              StorageControlSettings.newBuilder()
+                  .setCredentialsProvider(
+                      FixedCredentialsProvider.create(new CredentialAdapter(this.credential)))
```

Comment:
Did reusing credentials help improve first request latency?

Reply:
Which credential are you referring to? Do you recommend a different way of implementing this?

Reply:
Oh, for some reason i thought this was reusing credentials instances but that's not the case.

We have been talking offline about performance issue with GAPIC with initial request; I've been wondering if the same credential provider could help improve initial latency.